lib/storage/test_storage.ex

```diff
@@ -1,35 +1,114 @@
+          if test_environment?() do
+            {Uploadex.TestStorage, []}
+          else
+            {Uploadex.FileStorage, [...]}
+          end
+        end
+
+        @impl true
+        def accepted_extensions(%User{}, _field), do: ~w(.jpg .png)
+
+        defp test_environment?() do
+          # This env must be set in your config files depending on the environment.
+          Application.fetch_env!(:my_app, :environment) == :test
+        end
+      end
+
+  ## Using
+
+  In your `ExUnit` tests, add a `setup` block starting the storage:
+
+      setup do
+        Uploadex.TestStorage.start_link()
+        :ok
+      end
+
+  Then, in the tests, you can use the `get_stored/1`, `get_deleted/1` and `get_opts`:
+
+      test "some test with files" do
+        assert ["document-1.pdf", "document-2.pdf"] == Uploadex.TestStorage.get_stored()
+        assert ["deleted-document.pdf"] == Uploadex.TestStorage.get_deleted()
+        assert [] == Uploadex.TestStorage.get_opts()
+      end
+
+  This module is just the base for testing, check `Uploadex.Testing` for a more convenient way to test uploads.
   """
 
   use Agent
 
   @behaviour Uploadex.Storage
 
-  def start_link(initial_value \\ %{stored: [], deleted: [], opts: []}) do
-    Agent.start_link(fn -> initial_value end, name: __MODULE__)
+  def start_link(initial_state \\ %{}) when is_map(initial_state) do
+    agent_name = initial_state |> Map.get(:opts, []) |> get_agent_name()
+
+    default_state = %{stored: [], deleted: [], opts: [agent_name: agent_name]}
+    initial_state = Map.merge(default_state, Map.new(initial_state))
+
+    Agent.start_link(fn -> initial_state end, name: agent_name)
   end
 
-  def get_stored, do: Agent.get(__MODULE__, & &1.stored)
+  def get_stored(opts \\ []) do
+    opts
+    |> get_agent_name()
+    |> Agent.get(& &1.stored)
+  end
 
-  def get_deleted, do: Agent.get(__MODULE__, & &1.deleted)
+  def get_deleted(opts \\ []) do
+    opts
+    |> get_agent_name()
+    |> Agent.get(& &1.deleted)
+  end
 
-  def get_opts, do: Agent.get(__MODULE__, & &1.opts)
+  def get_opts(opts \\ []) do
+    opts
+    |> get_agent_name()
+    |> Agent.get(& &1.opts)
+  end
 
   @impl true
   def store(file, opts) do
-    Agent.update(__MODULE__, fn state -> Map.update!(state, :stored, & &1 ++ [file]) end)
-    Agent.update(__MODULE__, fn state -> Map.put(state, :opts, opts) end)
+    opts
+    |> get_agent_name()
+    |> Agent.update(fn state -> Map.update!(state, :stored, &(&1 ++ [file])) end)
+
+    opts
+    |> get_agent_name()
+    |> Agent.update(fn state -> Map.update!(state, :opts, &Keyword.merge(&1, opts)) end)
+
     :ok
   end
 
   @impl true
   def delete(file, opts) do
-    Agent.update(__MODULE__, fn state -> Map.update!(state, :deleted, & &1 ++ [file]) end)
-    Agent.update(__MODULE__, fn state -> Map.put(state, :opts, opts) end)
+    opts
+    |> get_agent_name()
+    |> Agent.update(fn state -> Map.update!(state, :deleted, &(&1 ++ [file])) end)
```

Comment:
I guess we don't need 2 Agent.updates, right?

Could be slightly simpler if it was:
```elixir
Agent.update(fn state ->
  state
  |> Map.update!(:deleted, &(&1 ++ [file]))
  |> Map.update!(:opts, &Keyword.merge(&1, opts))
end)
```